src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,162 @@
+
+# Define LFRic symbols and datatypes
+modules = [{"name": "constants_mod", "vars": ["i_def", "r_def", "l_def"]}]
+
+# Generate LFRic module symbols from definitions
+for module in modules:
+    MODULE_NAME = module["name"]
+    exec("{0} = ContainerSymbol('{1}')\n".format(
+        MODULE_NAME.upper(), MODULE_NAME))
+    for module_var in module["vars"]:
+        exec("{0} = DataSymbol('{1}', DeferredType(), interface="
+             "GlobalInterface({2}))".format(
+                 module_var.upper(), module_var, MODULE_NAME.upper()))
+
+scalar_datatypes = [
+    {"name": "lfric integer scalar", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "integer_scalar"},
+    {"name": "lfric real scalar", "intrinsic": "real",
+     "precision": "r_def", "vars": [], "short_name": "real_scalar"},
+    {"name": "lfric logical scalar", "intrinsic": "boolean",
+     "precision": "l_def", "vars": [], "short_name": "logical_scalar"},
+    {"name": "cell position", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "mesh height", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "operator size", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "ncell_3d"},
+    {"name": "number of dofs", "intrinsic": "integer",
+     "precision": "i_def", "vars": ["fs"], "short_name": "ndf"},
+    {"name": "number of unique dofs", "intrinsic": "integer",
+     "precision": "i_def", "vars": ["fs"], "short_name": "undf"},
+    {"name": "number of faces", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nfaces"},
+    {"name": "number of edges", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nedges"},
+    {"name": "number of qr points in horizontal", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp_h"},
+    {"name": "number of qr points in vertical", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp_v"},
+    {"name": "number of qr points", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp"}]
+
+for info in scalar_datatypes:
+    NAME = "".join(info["name"].title().split())
+    INTRINSIC = info["intrinsic"].upper()
+    PRECISION = info["precision"]
+    ARGS = ["self", "name"] + info["vars"] + ["interface=None"]
+    VARS = ["        self.{0} = {0}".format(var) for var in info["vars"]]
+    exec('''
+class {0}DataType(ScalarType):
+    def __init__(self):
+        super({0}DataType, self).__init__(
+            ScalarType.Intrinsic.{1}, {2})
+    '''.format(NAME, INTRINSIC, PRECISION.upper()))
+    exec('''
+class {0}DataSymbol(DataSymbol):
+    def __init__({1}):
+{2}
+        super({0}DataSymbol, self).__init__(
+            name, {0}DataType(), interface=interface)
+    '''.format(NAME, ", ".join(ARGS), "\n".join(VARS)))
+
+field_datatypes = [
```

Comment:
It's not crucial but a named tuple might be better than a dictionary here and in similar places in this file. (Since the keys are known and are always the same.)

Reply:
Agreed and modified.